aten/src/ATen/core/TransformationHelper.h

```diff
@@ -74,4 +76,59 @@ C10_HOST_DEVICE inline dist_acctype<T> uniform_real(V val, T from, T to) {
   return (x * (to - from) + from);
 }
 
+/**
+ * Transforms normally distributed `val` with mean 0.0 and standard deviation 1.0 to
+ * normally distributed with `mean` and standard deviation `std`.
+ */
+template <typename T>
+C10_HOST_DEVICE inline T normal(T val, T mean, T std) {
+  return val * std + mean;
+}
+
+/**
+ * Transforms uniformly distributed `val` between 0.0 and 1.0 to
+ * Cauchy distribution with location parameter `median` and scale parameter `sigma`.
+ */
+template <typename T>
+C10_HOST_DEVICE inline T cauchy(T val, T median, T sigma) {
+  // https://en.wikipedia.org/wiki/Cauchy_distribution#Cumulative_distribution_function
+  return median + sigma * at::tan(static_cast<T>(M_PI) * (val - static_cast<T>(0.5)));
+}
+
+/**
+ * Transforms uniformly distributed `val` between 0.0 and 1.0 to
+ * exponentialy distributed with `lambda` parameter of the distribution.
+ */
+template <typename T>
+C10_HOST_DEVICE __ubsan_ignore_float_divide_by_zero__ inline T exponential(T val, T lambda) {
+  // https://en.wikipedia.org/wiki/Exponential_distribution#Generating_exponential_variates
+  // Different implementations for CUDA and CPU to preserve original logic
+  // TODO: must be investigated and unified!!!
+  // https://github.com/pytorch/pytorch/issues/38662
+#if defined(__CUDACC__) || defined(__HIPCC__)
+  return static_cast<T>(-1.0) / lambda * at::log(val);
+#else
+  return static_cast<T>(-1.0) / lambda * at::log(static_cast<T>(1.0) - val);
+#endif
+}
+
+/**
+ * Transforms uniformly distributed `val` between 0.0 and 1.0 to
+ * geometricaly distributed with success probability `p`.
+ */
+template <typename T>
+C10_HOST_DEVICE inline T geometric(T val, T p) {
+  // https://en.wikipedia.org/wiki/Geometric_distribution#Related_distributions
+  return static_cast<T>(::ceil(at::log(val) / at::log(static_cast<T>(1.0) - p)));
```

Comment:
I have two questions here. 

(1) Is `val` in `[0, 1)`? (I thought so based on `torch.rand`). If so, should we use `log(1- val)` instead? 
Similar to exponential case,
`log(1-val)` for `val` in `[0, 1)`
`log(val)` for `val` in `(0, 1]`

(2) Instead of `ceil(...)`, should we use `floor(...) + 1` ? If `val=0` (although extremely rare), then `ceil( log(1-val) ) = 0` then `x =0`, but `torch.Tensor.geometric_` variate excludes 0.  Instead, if we use `floor(...) + 1`, this is avoided.

In summary, should we use something like `floor( log(1-val) / log(1-p) ) + 1` instead?

Reply:
If I'm not mistaken, is the documentation here little off? https://pytorch.org/docs/stable/generated/torch.Tensor.geometric_.html

Currently,
```
f(X = k) = (p**(k-1)) * (1-p)
```
this equation is correct if `p` is the probability of failure. But typically, `p` is the probability of success and `q = 1 - p` is the probability of failure. 

So either (1) `p` notation is untypical (should be clarified that `p` denotes probability of failure), but the equation is correct; or (2) `p` notation is typical (denotes probability of success), but the equation is incorrect - should be `f(X = k) = ((1-p)**(k-1)) * p`.

I think it's the second case (2) though. 
```python
torch.mean(torch.randn(10000).geometric_(p=0.7))
```
gives close to `E[X] = 1/p` value, where `p` denotes the probability of success.